src/renderers/sm_ray_tracer.cpp

```diff
@@ -1,14 +1,207 @@
+	{
+		return { -1 };
+	}
+
+	MUU_PURE_GETTER
+	static hit_result select(const hit_result& a, const hit_result& b) noexcept
+	{
+		if (!a)
+			return b;
+
+		return !b || a.distance <= b.distance ? a : b;
+	}
+
+	using scatter_func = std::optional<ray> MUU_VECTORCALL(const rt::scene& scene,
+														   const ray& r,
+														   const hit_result& hit,
+														   vec3& attenuation) noexcept;
+
+	MUU_PURE_GETTER
+	static std::optional<ray> MUU_VECTORCALL lambert_scatter(const rt::scene& scene,
+															 const ray& r,
+															 const hit_result& hit,
+															 vec3& attenuation) noexcept
+	{
+		attenuation = vec3{ scene.materials.albedo()[hit.material] * scene.materials.reflectivity()[hit.material] };
+
+		auto scatter = hit.normal + random_unit_vector();
+		if (scatter.approx_zero())
+			scatter = hit.normal;
+		scatter.normalize();
+
+		return ray{ r.at(hit.distance), scatter };
+	}
+
+	MUU_PURE_GETTER
+	static std::optional<ray> MUU_VECTORCALL metal_scatter(const rt::scene& scene,
+														   const ray& r,
+														   const hit_result& hit,
+														   vec3& attenuation) noexcept
+	{
+		attenuation = vec3{ scene.materials.albedo()[hit.material] * scene.materials.reflectivity()[hit.material] };
+
+		vec3 scatter = reflect(vec3::normalize(r.direction), hit.normal)
+					 + scene.materials.roughness()[hit.material] * random_unit_vector();
+		if (vec3::dot(scatter, hit.normal) <= 0.0f)
+			return {};
+
+		scatter.normalize();
+		return ray{ r.at(hit.distance), scatter };
+	}
+
+	static constexpr auto scatter_funcs = []() noexcept
+	{
+		std::array<scatter_func*, magic_enum::enum_count<material_type>()> funcs{};
+
+		for (auto& func : funcs)
+			func = lambert_scatter; // default to lambert for unimplemented brdfs
+
+		funcs[muu::unwrap(material_type::metal)] = metal_scatter;
+
+		return funcs;
+	}();
+
+	[[nodiscard]]
+	static vec3 MUU_VECTORCALL trace(const rt::scene& scene, const ray r, unsigned max_bounces) noexcept
+	{
+		if (!(max_bounces--))
+			return {};
+
+		auto hit = test_planes(scene, r);
+		hit		 = select(test_spheres(scene, r), hit);
+		hit		 = select(test_boxes(scene, r), hit);
+		if (!hit)
+			return vec3::lerp(colours::white.rgb, vec3{ 0.5f, 0.7f, 1.0f }, 0.5f * (r.direction.y + 1.0f));
+
+		vec3 attenuation;
+		if (const auto scatter = scatter_funcs[static_cast<size_t>(scene.materials.type()[hit.material])](scene, //
+																										  r,
+																										  hit,
+																										  attenuation))
+			return attenuation * trace(scene, *scatter, max_bounces);
+
+		return {};
+	}
+
 	struct sm_ray_tracer final : renderer_interface
 	{
-		void render(const rt::scene& /*scene*/, image_view& /*pixels*/, muu::thread_pool& /*threads*/) noexcept override
+		void render(const rt::scene& scene /*scene*/,
+					image_view& pxls /*pixels*/,
+					muu::thread_pool& threads /*threads*/) noexcept override
```

Comment:
These inline comments were in the original because the arguments were unused. They are used now, so the comments can go.

```suggestion
		void render(const rt::scene& scene,
					image_view& pxls,
					muu::thread_pool& threads) noexcept override
```